lwr:
  # Publish all joint states -----------------------------------
  joint_state_controller:
    type: joint_state_controller/JointStateController
    publish_rate: 100  

  # Position Controllers ---------------------------------------
  joint0_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_0_joint
    pid: {p: 21000.0, i: 10.0, d: 0.0}
  joint1_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_1_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}
  joint2_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_2_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}
  joint3_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_3_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}
  joint4_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_4_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}
  joint5_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_5_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}
  joint6_position_controller:
    type: effort_controllers/JointPositionController
    joint: lwr_6_joint
    pid: {p: 50000.0, i: 10.0, d: 0.0}

  # Trajectory Controllers ---------------------------------------
  joint_trajectory_controller:
    type: effort_controllers/JointTrajectoryController
    joints: 
      - lwr_0_joint
      - lwr_1_joint
      - lwr_2_joint
      - lwr_3_joint
      - lwr_4_joint
      - lwr_5_joint
      - lwr_6_joint
    gains:
      lwr_0_joint: {p: 21000,  d: 0, i: 10, i_clamp: 30}
      lwr_1_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}
      lwr_2_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}
      lwr_3_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}
      lwr_4_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}
      lwr_5_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}
      lwr_6_joint: {p: 50000,  d: 0, i: 10, i_clamp: 30}

    constraints:
      goal_time: 0.5                   # Override default

    state_publish_rate:  25            # Override default
    action_monitor_rate: 30            # Override default
    stop_trajectory_duration: 0        # Override default

  #   Joint Impedance Controllers
  JointImpedanceControl:
    type: lwr_controllers/JointImpedanceController
    root_name: world
    tip_name: lwr_7_link
    stiffness_gains: 300
    damping_gains: .7

  # Inverse Dynamics Controllers
  InverseDynamicsControl:
    type: lwr_controllers/InverseDynamicsController
    root_name: world
    tip_name: lwr_7_link

  # Computed Torque Controllers
  ComputedTorqueControl: 
    type: lwr_controllers/ComputedTorqueController
    root_name: world
    tip_name: lwr_7_link

  # One Task Inverse Kinematics
  OneTaskInverseKinematics: 
    type: lwr_controllers/OneTaskInverseKinematics
    root_name: world
    tip_name: lwr_7_link
    pid_lwr_0_joint: {p: 250,  i: 10, d: 30, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_1_joint: {p: 220,  i: 10, d: 30, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_2_joint: {p: 150,  i: 6, d: 20, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_3_joint: {p: 150,  i: 5, d: 12, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_4_joint: {p: 90,  i: 5, d: 10, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_5_joint: {p: 40,  i: 5, d: 7, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_6_joint: {p: 15,  i: 2, d: 5, i_clamp_min: -0.3, i_clamp_max: 0.3}

  # Multi Task Priority Inverse Kinematics
  MultiTaskPriorityInverseKinematics: 
    type: lwr_controllers/MultiTaskPriorityInverseKinematics
    root_name: world
    tip_name: lwr_7_link
    pid_lwr_0_joint: {p: 250,  i: 10, d: 30, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_1_joint: {p: 220,  i: 10, d: 30, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_2_joint: {p: 150,  i: 6, d: 20, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_3_joint: {p: 150,  i: 5, d: 12, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_4_joint: {p: 90,  i: 5, d: 10, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_5_joint: {p: 40,  i: 5, d: 7, i_clamp_min: -0.3, i_clamp_max: 0.3}
    pid_lwr_6_joint: {p: 15,  i: 2, d: 5, i_clamp_min: -0.3, i_clamp_max: 0.3}

  # One Task Inverse Dynamics JL
  OneTaskInverseDynamicsJL: 
    type: lwr_controllers/OneTaskInverseDynamicsJL
    root_name: world
    tip_name: lwr_7_link

  # Multi Task Priority Inverse Dynamics
  MultiTaskPriorityInverseDynamics: 
    type: lwr_controllers/MultiTaskPriorityInverseDynamics
    root_name: world
    tip_name: lwr_7_link

  # Minimum Effort Inverse Dynamics
  MinimumEffortInverseDynamics: 
    type: lwr_controllers/MinimumEffortInverseDynamics
    root_name: world
    tip_name: lwr_7_link

  # Backstepping Controller
  BacksteppingController: 
    type: lwr_controllers/BacksteppingController
    root_name: world
    tip_name: lwr_7_link

  # Dynamic Sliding Mode Controller
  DynamicSlidingModeController: 
     type: lwr_controllers/DynamicSlidingModeController
     root_name: world
     tip_name: lwr_7_link

  # Dynamic Sliding Mode Controller Task Space
  DynamicSlidingModeControllerTaskSpace: 
     type: lwr_controllers/DynamicSlidingModeControllerTaskSpace
     root_name: world
     tip_name: lwr_7_link
